### PR TITLE
[AsmPrinter] Reduce AsmPrinterHandlers virt. fn calls

### PR DESCRIPTION
**Pull request by aengelke**
Currently, an AsmPrinterHandler has several methods that allow to dynamically hook in unwind or debug info emission, e.g. at begin/end of every function or instruction. The class hierarchy and the actually overriden functions are as follows:

    (SymSz=setSymbolSize, mFE=markFunctionEnd, BBS=BasicBlockSection,
     FL=Funclet; b=beginX, e=endX)
                          SymSz   Mod Fn  mFE BBS FL  Inst
    AsmPrinterHandler     -       -   -   -   -   -   -
    ` PseudoProbeHandler  -       -   -   -   -   -   -
    ` WinCFGuard          -       e   e   -   -   -   -
    ` EHStreamer          -       -   -   -   -   -   -
      ` DwarfCFIException -       e   be  -   be  -   -
      ` ARMException      -       -   be  e   -   -   -
      ` AIXException      -       -   e   -   -   -   -
      ` WinException      -       e   be  e   -   be  -
      ` WasmException     -       e   be  -   -   -   -
    ` DebugHandlerBase    -       b   be  -   be  -   be
      ` BTFDebug          -       e   -   -   -   -   b
      ` CodeViewDebug     -       be  -   -   -   -   b
      ` DWARFDebug        yes     be  -   -   -   -   b

Doing virtual function calls per instruction is costly and useless when the called function does nothing.

This commit performs the following clean-up/improvements:

- PseudoProbeHandler is no longer an AsmPrinterHandler -- it used nothing of its functionality to hook in at the possible points. This avoids virtual function calls when a pseudo probe printer is present.

- DebugHandlerBase is no longer an AsmPrinterHandler, but a separate base class. DebugHandlerBase is the only remaining "hook" for begin/end instruction and setSymbolSize (only used by DWARFDebug). begin/end for function and basic block sections are never overriden and therefore are no longer virtual. (Originally I intended there to be only one debug handler, but BPF as the only target supports two at the same time: DWARF and BTF.)

- AsmPrinterHandler no longer has begin/end instruction and setSymbolSize hooks -- these were only used by DebugHandlerBase. This avoid iterating over handlers in every instruction.

- Remove NamedRegionTimer from instruction loop. Checking a global variable for every instruction (and doing an out-of-line function call) is too expensive for a profiling functionality.

```
    AsmPrinterHandler     Mod Fn  mFE BBS FL
    ` WinCFGuard          e   e   -   -   -
    ` EHStreamer          -   -   -   -   -
      ` DwarfCFIException e   be  -   be  -
      ` ARMException      -   be  e   -   -
      ` AIXException      -   e   -   -   -
      ` WinException      e   be  e   -   be
      ` WasmException     e   be  -   -   -

                          SymSz   Mod Fn  BBS Inst
    DebugHandlerBase      -       b   be  be  be
    ` BTFDebug            -       e           b
    ` CodeViewDebug       -       be          b
    ` DWARFDebug          yes     be          b

    PseudoProbeHandler (no shared methods)
```

This results in a performance improvement, especially in the -O0 -g0 case with unwind information (e.g., JIT baseline).

---

http://llvm-compile-time-tracker.com/compare.php?from=54cb5ca9f48fc542b920662a0eee7c0e6f35bee0&to=9ce05be2903285e1681ced36e5862ad3e155cb17&stat=instructions%3Au

### CODE REVIEW

**Comment by aengelke** [comment]
NB: supporting multiple debug handlers slightly regresses performance over my initial approach of zero or one debug handlers ([c-t-t](http://llvm-compile-time-tracker.com/compare.php?from=be4a0b21bf3b146a75759ec9b3cfaf8367eef16d&to=9ce05be2903285e1681ced36e5862ad3e155cb17&stat=instructions:u)). I also don't know if DWARF is actually useful for BPF or whether it is just there because it happened to work.

**Review by dwblaikie — COMMENTED** [review]
@aeubanks - any opinions on the compile time improvement here? (worth it? the complexity change isn't huge)
@JDevlieghere mabe you/someone else with debug info interest might have some opinions on how this changes the debug info handler interface - but it's pretty low impact

**Comment by aeubanks** [comment]
is the performance benefit from removing the timers or from the virtual call optimization? it would be nice to separate those two changes. I'm not sure if anyone is relying on the timers for profiling

**Comment by aengelke** [comment]
Removing timers is [very beneficial](http://llvm-compile-time-tracker.com/compare.php?from=54cb5ca9f48fc542b920662a0eee7c0e6f35bee0&to=84cac784134ba4a89a6596323df2c5304857e433&stat=instructions:u) (84ca), removing the virtual function calls has a [minor, but still postive effect](http://llvm-compile-time-tracker.com/compare.php?from=84cac784134ba4a89a6596323df2c5304857e433&to=46aab9a29b5ee06ed63c79f35bf9413827b28ee5&stat=instructions:u) on the instruction count (46aa).

```
Commit      stage1-O3        stage1-ReleaseThinLTO   stage1-ReleaseLTO-g     stage1-O0-g     stage2-O3        stage2-O0-g
46aab9a29b  62542M (-0.02%)  80380M (+0.00%)         93660M (+0.01%)         19172M (-0.04%) 55318M (-0.00%)  16856M (+0.01%)
84cac78413  62556M (-0.02%)  80379M (-0.02%)         93650M (-0.10%)         19181M (-0.26%) 55318M (+0.00%)  16854M (-0.18%)
54cb5ca9f4  62570M           80398M                  93744M                  19230M          55316M           16884M
```

**Comment by aeubanks** [comment]
Apologies for asking for extra work, but it would be good to do some git archaeology to see if these timers were intentionally added for someone to use (in which case we should ask around before removing them), or if they were randomly added without much thought

**Comment by aengelke** [comment]
e8dd2847ff3eb orginally introduced Dwarf timers in 2009. 6e6d1b2481b13 splitted timers between debug info and exception writing in 2009. f8dba24a9bba7 moved timers from DwarfDebug to AsmPrinter in 2010. It also introduced timers per instruction (back then called begin/endScope), which is currently the main performance problem. The commit message gives no reason for this. Later, timers were refactored multiple times and kept (NB: 1cd1444449cb6 introduced AsmPrinterHandler, generalizing things).

I think the timers per instruction are just there, because someone added them. I don't know how much use these timers see, because apart from a few refactoring, they didn't seem to get many changes.

File: llvm/include/llvm/CodeGen/DebugHandlerBase.h
Comment (MaskRay) [inline llvm/include/llvm/CodeGen/DebugHandlerBase.h:126]: `{};` => `{}`

**Review by MaskRay — COMMENTED** [review]
> PseudoProbeHandler is no longer an AsmPrinterHandler -- it used nothing of its functionality to hook in at the possible points. This avoids virtual function calls when a pseudo probe printer is present.

LGTM. cleanup after https://reviews.llvm.org/D91878

> DebugHandlerBase is no longer an AsmPrinterHandler, but a separate base class. DebugHandlerBase is the only remaining "hook" for begin/end instruction and setSymbolSize (only used by DWARFDebug). begin/end for function and basic block sections are never overriden and therefore are no longer virtual. (Originally I intended there to be only one debug handler, but BPF as the only target supports two at the same time: DWARF and BTF.)

LGTM. This part of this patch yields the largest win.

> AsmPrinterHandler no longer has begin/end instruction and setSymbolSize hooks -- these were only used by DebugHandlerBase. This avoid iterating over handlers in every instruction.

LGTM.

> Remove NamedRegionTimer from instruction loop. Checking a global variable for every instruction (and doing an out-of-line function call) is too expensive for a profiling functionality.

LGTM. Agreed that the overhead overweighs the potential use.
To address @aeubanks's comment, this could be split to a separate PR.

File: llvm/include/llvm/CodeGen/AsmPrinter.h
Comment (MaskRay) [inline llvm/include/llvm/CodeGen/AsmPrinter.h:209]: Personal note: `HandlerInfo` was made public by https://reviews.llvm.org/D74158 due to Julia's use. @vtjnash

Comment (aengelke) [reply llvm/include/llvm/CodeGen/AsmPrinter.h]: Good point, I added an addDebugHandler method so that the Julia use case should continue to work.

**Review by MaskRay — APPROVED** [review]
Thanks for the update.

**Review by aeubanks — APPROVED** [review]

File: llvm/unittests/CodeGen/AsmPrinterDwarfTest.cpp
Comment (vtjnash) [inline llvm/unittests/CodeGen/AsmPrinterDwarfTest.cpp:390]: Thanks @MaskRay. I am concerned that this is removing this particular API that is used by Julia to add a new DebugHandler. I am a mild NAK on doing so merely in the interest in reducing the internal instruction counter metric when it seems that you reported it doesn't actually improve performance. Refactoring is fine, and I don't mind if we need to change names of the API and such, as long as the functionality remains accessible to add beginInstruction hooks.

File: llvm/include/llvm/CodeGen/AsmPrinter.h
Comment (nikic) [inline llvm/include/llvm/CodeGen/AsmPrinter.h:23]: Can this be avoided somehow? It pulls the entire kitchen sink into AsmPrinter.

File: llvm/unittests/CodeGen/AsmPrinterDwarfTest.cpp
Comment (aengelke) [reply llvm/unittests/CodeGen/AsmPrinterDwarfTest.cpp:390]: For this purpose, I already added `addDebugHandler` where you can add a DebugHandler(Base), which still allows to hook in at every instruction. Could you check whether this replacement suits your needs?

> reducing the internal instruction counter metric

That's one way to phrase it -- improving performance of every single compilation done by LLVM would be mine. :shrug:

File: llvm/include/llvm/CodeGen/AsmPrinter.h
Comment (aengelke) [reply llvm/include/llvm/CodeGen/AsmPrinter.h:23]: Right now, I don't think so, HandlerInfo is templated and unique_ptr requires the definition for the destructor. After #97046, which removes HandlerInfo, it should be possible with an out-of-line destructor of AsmPrinter. I'll check next week (I won't merge before then anyway) -- thanks for pointing this out.

File: llvm/unittests/CodeGen/AsmPrinterDwarfTest.cpp
Comment (vtjnash) [reply llvm/unittests/CodeGen/AsmPrinterDwarfTest.cpp:390]: Instruction count may only have a weak correlation with wall time though, since instructions take vastly different times to retire, while your comment above seemed to show that this PR made everything measured at 0.5% slower (https://github.com/llvm/llvm-project/pull/96785#issuecomment-2193989698) if I am reading that right. I assume that is well within the noise factor, so not a blocker from me.

`addDebugHandler` looks good for me, but might need a new test here?

Comment (aengelke) [reply llvm/unittests/CodeGen/AsmPrinterDwarfTest.cpp:390]: > Instruction count may only have a weak correlation with wall time though

I know and also do time measurements locally. Indirect function calls are frequently less efficient than, e.g., arithmetic instructions.

> addDebugHandler looks good for me, but might need a new test here?

Good, thanks for checking! I'll add a test case.

Comment (aengelke) [reply llvm/unittests/CodeGen/AsmPrinterDwarfTest.cpp:390]: I added a test for addDebugHandler. If Julia's use case turns out to have significant problems after merging this, we can still adjust DebugHandlerBase afterwards to be more flexible.

**Pull request merged**
